File: PTA/1012.cpp
#include<iostream>
#include<stdio.h>
#include<vector>
using namespace std;
// scores[2000]里面存的是所有的id
vector<int> scores[2001];
vector<int> ranks[2000]; 
// 记录所有的id的下标
int ids[1000000];
double avg[2000];

int N,M;

void rankMethod(int id){
	int rank,tempIndex;
	// 取出下标 
	int index = ids[id];
	// 拿到平均值 
	double avgNum = avg[index];
	// 先计算平均值的排名,初始值为1，若有比他大的就排名加一 
	rank = 1;
	for(int temp:scores[2000]){
		//对每一个id，都比较其平均值
		if(avg[ids[temp]]>avgNum) rank++;
	}
	ranks[index].push_back(rank);
	
	// 接下来依次计算C、M、E的排名
	for(int i=0;i<3;i++){
		// 先将排名置为1 
		rank = 1; 
		for(int temp:scores[2000]){
		   //取出下标,取出对应的分数做比较，若大于当前id的分数，则排名+1 
		   tempIndex = ids[temp];
		   if(scores[tempIndex][i]>scores[index][i]) rank++;
	   }
	   ranks[index].push_back(rank);
	}
}

int main(){
	int Cs,Math,Eng;
	int tempId,tempIndex;
	cin>>N>>M;
	for(int i=0;i<1000000;i++) ids[i] = -1;
	char result[4];
	result[0] = 'A';
	result[1] = 'C';
	result[2] = 'M';
	result[3] = 'E';
	for(int i=0;i<N;i++){
		//scanf("%d",&tempId);
		cin>>tempId;
	    //scanf("%d %d %d",&Cs,&Math,&Eng);
	    cin>>Cs>>Math>>Eng;
	    // 计算平均值
	    avg[i] = (Cs+Math+Eng)/3.0;
	    scores[i].push_back(Cs);
	    scores[i].push_back(Math);
	    scores[i].push_back(Eng);
	    ids[tempId] = i;
	    scores[2000].push_back(tempId);
	}
	
	for(int i=0;i<M;i++){
		cin>>tempId;
		tempIndex = ids[tempId];
		if(tempIndex == -1){
			// id不存在 
			cout<<"N/A"<<endl;
		}else{
			// id存在，将此id下的所有成绩计算排名 
			rankMethod(tempId);
			int min = 0;
			for(int j=1;j<4;j++){
				if(ranks[tempIndex][j] < ranks[tempIndex][min]){
					min = j;
				}
			}
			cout<<ranks[tempIndex][min]<<" "<<result[min]<<endl;
		}
	}	
	return 0;
}
 

File: PTA/1012.java
package com.pat;

import java.util.List;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

public class Problem1012 {
	private static Map<String, List<Integer>> database = new HashMap<>();
	//rankMap 记录每个id的每一科的排名
	private static Map<String, List<Integer>> rankMap = new HashMap<>();
	// 每个id的list集合里的值依次是 C M E的分数 各科的排名(C,M,E,A)
	private static List<String> search = new ArrayList<>();
	// 记录每个id的所有科目平均分
	private static double[] avg = new double[2000];
	public static void rankMethod(String id) {
		int rank;
		List<Integer> scores = database.get(id);
		List<Integer> tempList = new ArrayList<>();
		
		// 计算平均值排名
		int index = scores.get(3);
	    double avgNum = avg[index];
		rank = 1;
		// 计算平均值的排名
	    for(String tempId:database.keySet()) {
			// 拿到某一科的成绩
		    int tempIndex = database.get(tempId).get(3);
		    if (avg[tempIndex] > avgNum) {
				rank++;
			}
	    }
	    tempList.add(rank);
		
		for (int i = 0; i < 3; i++) {
			//这里的每次循环都将id的某一科目的排名计算出来
			//计算C、M、E的排名
			int score = scores.get(i);
			//一开始排名第1，若有一个比他大，就加一
		    rank = 1;
			for(String tempId:database.keySet()) {
				// 拿到某一科的成绩
			    int tempScore = database.get(tempId).get(i);
			    if (tempScore > score) {
					rank++;
				}
		    }
			tempList.add(rank);
		}
		rankMap.put(id, tempList);
	}
	
	public static void main(String[]args) {
		Scanner scanner = new Scanner(System.in);
	    int N,M,Cs,Math,Eng;
	    String tempId;
	    char[] result = new char[4];
	    result[0] = 'A';
	    result[1] = 'C';
	    result[2] = 'M';
	    result[3] = 'E';
	    N = scanner.nextInt();
	    M = scanner.nextInt();
	    for(int i=0;i<N;i++){
	    	List<Integer> temp = new ArrayList<>();
	    	tempId = scanner.next();
	    	// 输入分数
	    	Cs = scanner.nextInt();
	    	Math = scanner.nextInt();
	    	Eng = scanner.nextInt();
	    	temp.add(Cs);
	    	temp.add(Math);
	    	temp.add(Eng);
	    	// 最后一个值是他的下标
	    	temp.add(i);
	    	// 计算平均值
	    	avg[i] = (Cs+Math+Eng)/3.0;
	    	database.put(tempId, temp);
	    }
	    
	    //System.out.println(database.toString());
	    for(int i=0;i<M;i++) {
	    	tempId = scanner.next();
	    	search.add(tempId);
	    }
	    
	    for(String id:search) {
	    	if(!database.containsKey(id)) {
	    		// 不存在此id
	    		System.out.println("N/A");
	    	}else {
	    		// 存在此id,则先计算排名
		    	rankMethod(id);
		    	int min = 0;
		    	List<Integer> rankList = rankMap.get(id);
		    	for (int i = 1; i < 4; i++) {
					if (rankList.get(i)<rankList.get(min)) {
						min =  i;
					}
				}
		    	System.out.println(rankList.get(min) + " " + result[min]);
	    	}
	    }
	}
}
